pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -0,0 +1,183 @@
+func NewManager(topologyPolicyName string) Manager {
+	klog.Infof("[topologymanager] Creating topology manager with %s policy", topologyPolicyName)
+	var policy Policy
+
+	switch policyName(topologyPolicyName) {
+
+	case PolicyPreferred:
+		policy = NewPreferredPolicy()
+
+	case PolicyStrict:
+		policy = NewStrictPolicy()
+
+	default:
+		klog.Errorf("[topologymanager] Unknow policy %s, using default policy %s", topologyPolicyName, PolicyPreferred)
+		policy = NewPreferredPolicy()
+	}
+
+	var hp []HintProvider
+	pnh := make(map[string]containers)
+	pm := make(map[string]string)
+	manager := &manager{
+		hintProviders:    hp,
+		podTopologyHints: pnh,
+		podMap:           pm,
+		policy:           policy,
+	}
+
+	return manager
+}
+
+func (m *manager) GetAffinity(podUID string, containerName string) TopologyHints {
+	return m.podTopologyHints[podUID][containerName]
+}
+
+func (m *manager) calculateTopologyAffinity(pod v1.Pod, container v1.Container) TopologyHints {
+	socketMask := socketmask.NewSocketMask(nil)
+	var maskHolder []string
+	count := 0
+	affinity := true
+	for _, hp := range m.hintProviders {
+		topologyHints := hp.GetTopologyHints(pod, container)
+		if topologyHints.Affinity && topologyHints.SocketAffinity != nil {
+			klog.Infof("[topologymanager] Topology Affinity.")
+			socketMask, maskHolder = socketMask.GetSocketMask(topologyHints.SocketAffinity, maskHolder, count)
+			count++
+		} else if topologyHints.Affinity && topologyHints.SocketAffinity == nil {
+			klog.Infof("[topologymanager] NO Topology Affinity.")
+			return TopologyHints{
+				SocketAffinity: []socketmask.SocketMask{socketMask},
+				Affinity:       false,
+			}
+		} else if !topologyHints.Affinity && topologyHints.SocketAffinity != nil {
+			klog.Infof("[topologymanager] Cross Socket Topology Affinity")
+			affinity = false
+			socketMask, maskHolder = socketMask.GetSocketMask(topologyHints.SocketAffinity, maskHolder, count)
+			count++
+		}
+	}
+	return TopologyHints{
+		SocketAffinity: []socketmask.SocketMask{socketMask},
+		Affinity:       affinity,
+	}
+}
+
+func (m *manager) AddHintProvider(h HintProvider) {
+	m.hintProviders = append(m.hintProviders, h)
+}
+
+func (m *manager) AddPod(pod *v1.Pod, containerID string) error {
+	m.podMap[containerID] = string(pod.UID)
+	return nil
+}
+
+func (m *manager) RemovePod(containerID string) error {
+	podUIDString := m.podMap[containerID]
+	delete(m.podTopologyHints, podUIDString)
+	delete(m.podMap, containerID)
+	klog.Infof("[topologymanager] RemovePod - Container ID: %v podTopologyHints: %v", containerID, m.podTopologyHints)
+	return nil
+}
+
+func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	klog.Infof("[topologymanager] Topology Admit Handler")
+	pod := attrs.Pod
+	c := make(containers)
+	klog.Infof("[topologymanager] Pod QoS Level: %v", pod.Status.QOSClass)
+
+	qosClass := pod.Status.QOSClass
+
+	if qosClass == "Guaranteed" {
+		for _, container := range pod.Spec.Containers {
```

Comment:
Is there a reason we don't gather TopologyHints for Init Containers?

In the `devicemanager`, devices allocated to Init Containers are reused by Spec Containers once the Init Containers finish.  If an Init Container doesn't have the opportunity to give any Topology Hints, is it possible that this could negatively impact the hints given by the Spec Container since (at least some of) the devices it will be allocated are already predetermined?

Reply:
Good point, I'll update the PR to the init containers into account as well